### ISSUE TITLE
Speed up namespaced_stored_object

### ISSUE THREAD

**Issue by dave-2** [catapult-project/catapult#3834]
`namespaced_stored_object.Get()` is exceedingly slow. A call takes .08 ± .06 seconds. If you do 100 of them in a loop, the user is waiting for 8 seconds. This is pretty easy to do, e.g. looking up the repository URL for every commit in a list of Pinpoint jobs.

If the user is logged out, the time is about 70% IP whitelist lookup and 30% retrieving the stored object. If the user is logged in, there's no IP whitelist check.

Storing dashboard configurations in a lightweight key-value store that just contains a single `JSONProperty` would make this take microseconds in most cases, thanks to the automatic in-context cache. Using `JSONProperty` would eliminate the pickling overhead. And using a single entity would avoid the additional complexity of `MultipartEntity`, since none of the configurations are over 1 MB, including the IP whitelist (29 KB) and the perf sheriff config (60 KB).

@eakuefner @benshayden @perezju

**Comment by dave-2** [comment]
A sample fast key-value store would be something like this:
```
from google.appengine.ext import ndb

from dashboard.common import datastore_hooks


def Put(key, value):
  Configuration(value=value, id=key).put()


def Get(key):
  return Configuration.get_by_id(key).value


def PutNamespaced(key, value, namespace=None):
  namespace = namespace or datastore_hooks.GetNamespace()
  Configuration(value=value, id=_NamespacedKey(namespace, key)).put()


def GetNamespaced(key, namespace=None):
  namespace = namespace or datastore_hooks.GetNamespace()
  return Configuration.get_by_id(_NamespacedKey(namespace, key)).value


class Configuration(ndb.Model):

  value = ndb.JsonProperty(required=True)


def _NamespacedKey(namespace, key):
  return '\n'.join((namespace, key))
```

**Comment by perezju** [comment]
That sgtm

**Comment by simonhatch** [comment]
In the CL reply, you mentioned TLS behaviour was weird, can you go into more detail about that?

**Comment by dave-2** [comment]
I'm sorry, I don't remember, and have already deleted the local branch where I tried it.
Regardless, implementing the local caching is significantly more code and has significantly more gotchas than speeding up the DB fetch. (E.g. stale data is a concern, thread safety)

**Comment by dave-2** [comment]
I did some archaeology and found some speedups we can do in `stored_object` without having to do any immediate data migration.

tl;dr: We can reduce the number of datastore fetches from 17 to 1 by fixing a small bug, and reduce warm hits from ms to μs by removing the memcache stuff.

There's a bug in `stored_object` where it always fetches `_MAX_NUM_PARTS` `PartEntities` . When serializing the data, it pads the values with `None` values.
```python
  for i in xrange(0, length, _CHUNK_SIZE):
    values.append(serialized[i:i + _CHUNK_SIZE])
  for i in xrange(len(values), _MAX_NUM_PARTS):
    values.append(None)
```
While these are later removed for the `put_multi`, the size is still used to calculate the number of parts.
```python
    num_parts = len(serialized_parts)
    for i in xrange(num_parts):
      if serialized_parts[i] is not None:
        part = PartEntity(id=i + 1, parent=self.key, value=serialized_parts[i])
        part_list.append(part)
    self.size = num_parts
```
In `GetPartsAsync `, it fetches all 16 parts.
```python
    part_keys = [ndb.Key(MultipartEntity, string_id, PartEntity, i + 1)
                 for i in xrange(self.size)]
    part_entities = yield ndb.get_multi_async(part_keys)
```
If we fix this bug, we can save 15 of these fetches for most objects, bringing the number of fetches from 17 to 2.

We can go even further than this: in `stored_object.py`, the maximum chunk size is 1,000,000 bytes. In reality, the maximum property size for NDB is 1,048,487 byes. That means we can fit an entire chunk inside the `MultipartEntity` object and avoid having to fetch any `PartEntities`. So we can do the entire thing in 1 fetch. Adding a field doesn't require a data migration.

I also believe that removing the `memcache` caching layer will speed up warm results (from milliseconds to microseconds) because we can make use of [NDB's automatic in-context cache](https://cloud.google.com/appengine/docs/standard/python/ndb/cache#incontext). Using memcache directly will skip over this caching layer.

I also looked at the overhead of JSON vs Pickling. When I ran the numbers, cPickling in the binary format (`protocol=pickle.HIGHEST_PROTOCOL`) was slightly faster and produced slightly smaller results than using JSON with no spaces. So we can stick with Pickling.

**Comment by simonhatch** [comment]
Nice find! Any idea how much this reduces the object lookup time by?

**Comment by dave-2** [comment]
I ran some rough numbers in the App Engine dev console. Code is at the end of this comment.
Fetching the IP whitelist 1000 times (warm):

> Full optimization: 0.97592997551
> No memcache: 8.93426012993
> Control: 15.1427800655

I got a 15x speedup over the control.
I don't have cold numbers, as it seems like it's hard to disable or purge the caches.

As for how this looks in context, I deployed a dev version and annotated the times for the request and `stored_object`. I requested a group report page:

> https://dev-dtu-f765073c-dot-chromeperf.appspot.com/group_report?sid=02b17e8f7d8d6f0b4b9921b7a01f0eb7307d10ef0b19333be982ca55a2fe1f15`

There was 1 call to `/group_report`:
> 9.99s total
> 0.28s stored_object (cold)
> 0.21s stored_object (warm)

There were 8 calls to `/graph_json` (but note that these happen in parallel):
> ~41.7s total
> ~8.4s stored_object (cold)
> 0s stored_object (warm).

So we would save a few seconds loading these graphs.

Testing code:
```python
import cPickle
import json
import time

from google.appengine.ext import ndb

from dashboard.common import stored_object

ITERATIONS = 1000
KEY = 'ip_whitelist'


def main():
  # Validate correctness and warm caches.
  assert GetOptimized(KEY) == GetNoMemcache(KEY) == stored_object.Get(KEY)

  start_time = time.time()
  for _ in xrange(ITERATIONS):
    GetOptimized(KEY)
  print 'Full optimization:', time.time() - start_time

  start_time = time.time()
  for _ in xrange(ITERATIONS):
    GetNoMemcache(KEY)
  print 'No memcache:', time.time() - start_time

  start_time = time.time()
  for _ in xrange(ITERATIONS):
    stored_object.Get(KEY)
  print 'Control:', time.time() - start_time


@ndb.synctasklet
def GetNoMemcache(key):
  result = yield GetAsyncNoMemcache(key)
  raise ndb.Return(result)


@ndb.tasklet
def GetAsyncNoMemcache(key):
  get_future = stored_object._GetValueFromDatastore(key)
  yield get_future
  raise ndb.Return(get_future.get_result())


def GetOptimized(key):
  key =  ndb.Key(stored_object.MultipartEntity, KEY,
                 stored_object.PartEntity, 1)
  value = key.get().value
  return cPickle.loads(value)


main()
```

**Comment by dave-2** [comment]
CL is up. https://codereview.chromium.org/3009343002/

Here's a migration script. Based on my testing, it will take about 3 hours to migrate all 188,000 stored_objects.
```python
# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import json
import logging

from google.appengine.api import taskqueue
from google.appengine.datastore import datastore_query

from dashboard.common import request_handler
from dashboard.common import stored_object as stored_object_module


_BATCH_SIZE = 1000


class MigrateStoredObjectHandler(request_handler.RequestHandler):

  def get(self):
    taskqueue.add(url='/migrate_stored_object')

  def post(self):
    cursor = datastore_query.Cursor(urlsafe=self.request.get('cursor'))
    query = stored_object_module.MultipartEntity.query()
    stored_objects, next_cursor, more = query.fetch_page(
        _BATCH_SIZE, start_cursor=cursor)

    for stored_object in stored_objects:
      key = stored_object.key.id()
      value = stored_object_module.Get(key)
      stored_object_module.Set(key, value)

    params = {'cursor': next_cursor.urlsafe()}
    taskqueue.add(url='/migrate_stored_object', params=params)
```

**Comment by dave-2** [comment]
I ran a one-off migration for the Pinpoint repositories and the IP whitelist. I confirmed that when running this code in the new version, it creates a MultipartEntity with `size == 1`, and when running this code in the old version, it back-migrates to `size == 16`.

```python
from dashboard.common import stored_object

key = 'ip_whitelist'
value = stored_object.Get(key)
stored_object.Set(key, value)
print value == stored_object.Get(key)
```

**Comment by dave-2** [comment]
There are still more optimizations we can make, but they may require backwards incompatible data migrations. See discussion on this CL: https://codereview.chromium.org/3009343002/

**Comment by benshayden** [comment]
This looks good enough for now?
